File: samples.vb/Tutorial/Movies.WPF/MovieDetailWindow.Presenter.vb
Imports DevZest.Data.Presenters

Partial Class MovieDetailWindow
    Private NotInheritable Class Presenter
        Inherits DataPresenter(Of Movie)
        Private Const LABEL_FORMAT As String = "{0}:"

        Protected Overrides Sub BuildTemplate(builder As TemplateBuilder)
            Dim title = Entity.Title.BindToTextBox()
            Dim releaseDate = Entity.ReleaseDate.BindToDatePicker()
            Dim genre = Entity.Genre.BindToTextBox()
            Dim price = Entity.Price.BindToTextBox()

            builder.GridColumns("Auto", "*") _
                .GridRows("Auto", "Auto", "Auto", "Auto") _
                .AddBinding(0, 0, Entity.Title.BindToLabel(title, LABEL_FORMAT)) _
                .AddBinding(0, 1, Entity.ReleaseDate.BindToLabel(releaseDate, LABEL_FORMAT)) _
                .AddBinding(0, 2, Entity.Genre.BindToLabel(genre, LABEL_FORMAT)) _
                .AddBinding(0, 3, Entity.Price.BindToLabel(price, LABEL_FORMAT)) _
                .AddBinding(1, 0, title) _
                .AddBinding(1, 1, releaseDate) _
                .AddBinding(1, 2, genre) _
                .AddBinding(1, 3, price)
        End Sub

        Public ReadOnly Property ID() As Integer
            Get
                Return Entity.ID(0).Value
            End Get
        End Property

        Public ReadOnly Property IsNew() As Boolean
            Get
                Return ID < 1
            End Get
        End Property

        Public Function SaveToDbAsync() As Task
            If IsNew Then
                Return Application.ExecuteAsync(Function(db) db.Movie.InsertAsync(DataSet))
            Else
                Return Application.ExecuteAsync(Function(db) db.Movie.UpdateAsync(DataSet))
            End If
        End Function
    End Class
End Class

File: samples/Tutorial/Movies.WPF/MovieDetailWindow.Presenter.cs
using DevZest.Data.Presenters;
using System.Threading.Tasks;

namespace Movies.WPF
{
    partial class MovieDetailWindow
    {
        private sealed class Presenter : DataPresenter<Movie>
        {
            private const string LABEL_FORMAT = "{0}:";

            protected override void BuildTemplate(TemplateBuilder builder)
            {
                var title = _.Title.BindToTextBox();
                var releaseDate = _.ReleaseDate.BindToDatePicker();
                var genre = _.Genre.BindToTextBox();
                var price = _.Price.BindToTextBox();

                builder
                    .GridColumns("Auto", "*")
                    .GridRows("Auto", "Auto", "Auto", "Auto")
                    .AddBinding(0, 0, _.Title.BindToLabel(title, LABEL_FORMAT))
                    .AddBinding(0, 1, _.ReleaseDate.BindToLabel(releaseDate, LABEL_FORMAT))
                    .AddBinding(0, 2, _.Genre.BindToLabel(genre, LABEL_FORMAT))
                    .AddBinding(0, 3, _.Price.BindToLabel(price, LABEL_FORMAT))
                    .AddBinding(1, 0, title)
                    .AddBinding(1, 1, releaseDate)
                    .AddBinding(1, 2, genre)
                    .AddBinding(1, 3, price);
            }

            public int ID
            {
                get { return _.ID[0].Value; }
            }

            public bool IsNew
            {
                get { return ID < 1; }
            }

            public Task SaveToDbAsync()
            {
                if (IsNew)
                    return App.ExecuteAsync(db => db.Movie.InsertAsync(DataSet));
                else
                    return App.ExecuteAsync(db => db.Movie.UpdateAsync(DataSet));
            }
        }
    }
}
